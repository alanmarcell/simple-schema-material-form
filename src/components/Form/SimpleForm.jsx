import React from 'react';
import PropTypes from 'prop-types';
import { omit, pathOr, merge } from 'ramda';
import { compose, withStateHandlers, pure } from 'recompact';
import { Button, TextField } from 'material-ui';

const getChildProps = ({ doc, setDoc }) => Child => {
  const { fieldName } = Child.props;
  const value = pathOr('', [fieldName], doc);

  const childProps = {
    ...Child.props,
    setDoc,
    doc,
    onChange: e => setDoc({ [fieldName]: e.target.value }),
    value,
    label: fieldName,
  };

  return React.createElement(
    Child.type,
    childProps,
    null,
  );
};

export const TextInput = props => {
  const { doc, setDoc, fieldName } = props;
  const value = pathOr('', [fieldName], doc);
  const cProps = omit(['fieldName', 'setDoc', 'doc'], props);

  const inputProps = {
    ...cProps,
    onChange: e => setDoc({ [fieldName]: e.target.value }),
    value,
    label: fieldName,
  };
  return <TextField {...inputProps} />;
};

/** Our Simple Form Test Utility */
const SimpleForm = (props) => {
  const { doc, setDoc, onSubmit } = props;
  const setChildProps = getChildProps({ doc, setDoc });
  return (
    <form >
      {React.Children.map(props.children, setChildProps)}
      <Button onClick={() => onSubmit(doc)} >Submit</Button>
    </form>
  );
};

/** This is our base enhancer to handle doc changes */
const enhance = compose(
  pure,
  withStateHandlers(
    /** this our initial doc */
    { doc: {} },
    {
      /** this is the doc state handler */
      setDoc: ({ doc }) => value => {
        const merged = merge(doc, value);
        return {
          doc: merged,
        };
      },
    },
  ),
);

const EnhancedSimpleForm = enhance(SimpleForm);

SimpleForm.defaultProps = {
  /** This is our fields */
  // children: null,
};

SimpleForm.propTypes = {
  /** doc defines the form state */
  doc: PropTypes.object.isRequired,

  /** handle the form state */
  setDoc: PropTypes.func.isRequired,

  children: PropTypes.array.isRequired,
};

export default EnhancedSimpleForm;
