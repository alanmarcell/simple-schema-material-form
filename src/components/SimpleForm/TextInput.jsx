import React from 'react';
// import PropTypes from 'prop-types';
import { omit, pathOr } from 'ramda';
// import { compose, withStateHandlers, pure } from 'recompact';
import { TextField } from 'material-ui';


const TextInput = props => {
  const { doc, setDoc, fieldName } = props;
  const value = pathOr('', [fieldName], doc);
  const cProps = omit(['fieldName', 'setDoc', 'doc'], props);

  const inputProps = {
    ...cProps,
    onChange: e => setDoc({ [fieldName]: e.target.value }),
    value,
    label: fieldName,
  };
  return <TextField {...inputProps} />;
};

export default TextInput;
