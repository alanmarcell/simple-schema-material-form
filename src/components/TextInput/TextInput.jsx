import React from 'react';

const TextInput = _props => {
  return (<div>TextInput</div>);
};

export default TextInput;
