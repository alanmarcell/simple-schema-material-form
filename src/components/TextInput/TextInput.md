# TextField With Recompose:

```jsx
const Values = require('./Values').default;
props = { fieldName : 'alan', value: 'Initial Value' };
initialState = { value: props.value, fieldName: props.fieldName};
<div>
  <Values {...state} />
  <TextInput
    {...props}
    onChange={e => { setState({value: e.target.value}); props.updateStateValue(e.target.value)}}
    ref={ref =>  (props = ref)} />
  <TextInput fieldName="thais" />
</div>
```
