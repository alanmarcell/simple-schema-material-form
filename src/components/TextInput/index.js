import { compose, withState } from 'recompose';

import TextInputComponent from './TextInput';

const enhance = compose(withState('value', 'setvalue', ({ value }) => value));

export { TextInputComponent, enhance };

export default TextInputComponent;
