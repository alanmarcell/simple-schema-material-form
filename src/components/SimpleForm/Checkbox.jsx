import React from 'react';
import { merge } from 'ramda';
import MaterialCheckbox from 'material-ui/Checkbox';
import { FormControl, FormLabel, FormControlLabel, FormGroup } from 'material-ui/Form';

const Checkbox = props => {
  const {
    setDoc, label, selectOptions = [], fieldName, doc,
  } = props;

  return (
    <FormControl >
      <FormLabel>{label}</FormLabel>
      <FormGroup row>
        {selectOptions.map(i =>
          (<FormControlLabel
            key={i.key}
            control={<MaterialCheckbox
              checked={i.checked}
              onChange={(_event, val) => setDoc(merge(doc, { [fieldName]: { ...doc[fieldName], [i.key]: val } }))}
              value={i.key || ''}
              id={i.key}
            />}
            label={i.label}
          />))
        }
      </FormGroup>
    </FormControl >
  );
};

// const Checkbox = props => {
//   const inputProps = {
//     ...props,
//   };
//   return <CheckboxField selectOptions={[{ label: 1, key: 'um' }, { label: 2, key: 'dois' }]} {...inputProps} />;
// };

export default Checkbox;
