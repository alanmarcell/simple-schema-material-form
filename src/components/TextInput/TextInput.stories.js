import React from 'react';
import { storiesOf } from '@storybook/react';
import { withState } from '@dump247/storybook-state';
import withTests from '../../stories/withTests';
import TextInput from './';

storiesOf('Fields|TextInput', module)
  .addDecorator(withTests('TextInput'))
  .add('simple', withState({ doc: {} }, (store) => (
    <TextInput
      {...store.state}
      setDoc={e => store.set({ doc: e })}
      fieldName="World"
    />
  )));

