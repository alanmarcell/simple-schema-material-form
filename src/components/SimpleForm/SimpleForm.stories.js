import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import SimpleForm from './SimpleForm';
import TextInput from '../TextInput';
import Checkbox from './Checkbox';
import withTests from '../../stories/withTests';

storiesOf('SimpleForm', module)
  .addDecorator(withTests('SimpleForm'))
  .add('with TextInput', () => (
    <SimpleForm onSubmit={action('doc')} >
      <TextInput fieldName="test" />
    </SimpleForm>
  ))
  .add('with 2 TextInput', () => (
    <SimpleForm onSubmit={action('doc')} >
      <TextInput fieldName="test" />
      <TextInput fieldName="test2" />
    </SimpleForm>
  ))
  .add('with Checkbox', () => (
    <SimpleForm onSubmit={action('doc')} >
      <Checkbox selectOptions={[{ label: 1, key: 'um' }, { label: 2, key: 'dois' }]} fieldName="test" />
    </SimpleForm>
  ))
  .add('with Checkbox and Text Iput', () => (
    <SimpleForm onSubmit={action('doc')} >
      <Checkbox selectOptions={[{ label: 1, key: 'um' }, { label: 2, key: 'dois' }]} fieldName="test" />
      <TextInput fieldName="test2" />
    </SimpleForm>
  ));

