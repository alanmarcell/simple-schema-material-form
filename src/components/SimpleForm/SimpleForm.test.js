import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';
import { Button } from 'material-ui';
import SimpleForm from './SimpleForm';
import TextInput from './TextInput';

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<SimpleForm ><TextInput fieldName="test" /></SimpleForm>, div);
  ReactDOM.unmountComponentAtNode(div);
});


describe('Test within Form', () => {
  it('shallow a TextInput', () => {
    const onSubmit = jest.fn();
    const SimpleFormComponent = (
      <SimpleForm onSubmit={doc => onSubmit(doc)}>
        <TextInput fieldName="test" />
      </SimpleForm>);

    const component = shallow(SimpleFormComponent);
    const newValue = 'Type Sample';

    const { children } = component.props();

    const Children = shallow(children[0][0]);

    Children.simulate('change', {
      target: { value: newValue },
    });

    component.update();
    const newState = component.state();
    const newChildren = component.props().children[0][0];
    const NewChildren = shallow(newChildren);

    const { doc } = newState.childProps;
    const test2 = newState.childProps.doc.test;

    expect(test2).toBe(newValue);
    expect(NewChildren.props().value).toBe(newValue);
    expect(component).toMatchSnapshot();


    const button = component.find(Button);
    button.simulate('click');

    expect(onSubmit).toHaveBeenCalled();
    expect(onSubmit).toBeCalledWith(doc);
  });


  it('shallow a 2 TextInput', () => {
    const onSubmit = jest.fn();
    const SimpleFormComponent = (
      <SimpleForm onSubmit={doc => onSubmit(doc)}>
        <TextInput fieldName="test" />
        <TextInput fieldName="test2" />
      </SimpleForm>);

    const component = shallow(SimpleFormComponent);

    const newValue = 'Type Sample';
    const { children } = component.props();
    const TextInput1 = shallow(children[0][0]);
    const TextInput2 = shallow(children[0][1]);

    TextInput1.simulate('change', {
      target: { value: newValue },
    });
    TextInput2.simulate('change', {
      target: { value: newValue },
    });
    const newState = component.state();
    const { doc } = newState.childProps;

    component.update();

    const newChildren = component.props().children;
    const NewTextInput1 = shallow(newChildren[0][0]);
    const NewTextInput2 = shallow(newChildren[0][1]);

    const button = component.find(Button);
    button.simulate('click');

    const test2 = newState.childProps.doc.test;

    const expectedDoc = {
      test: newValue,
      test2: newValue,
    };

    expect(doc).toEqual(expectedDoc);

    expect(test2).toBe(newValue);
    expect(NewTextInput1.props().value).toBe(newValue);
    expect(NewTextInput2.props().value).toBe(newValue);
    expect(component).toMatchSnapshot();
    expect(onSubmit).toHaveBeenCalled();
    expect(onSubmit).toBeCalledWith(doc);
  });
});
