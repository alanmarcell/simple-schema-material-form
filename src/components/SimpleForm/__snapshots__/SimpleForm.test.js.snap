// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test within Form shallow a 2 TextInput 1`] = `
<form>
  <TextInput
    doc={
      Object {
        "test": "Type Sample",
        "test2": "Type Sample",
      }
    }
    fieldName="test"
    key=".0"
    label="test"
    setDoc={[Function]}
    value="Type Sample"
  />
  <TextInput
    doc={
      Object {
        "test": "Type Sample",
        "test2": "Type Sample",
      }
    }
    fieldName="test2"
    key=".1"
    label="test2"
    setDoc={[Function]}
    value="Type Sample"
  />
  <WithStyles(Button)
    onClick={[Function]}
  >
    Submit
  </WithStyles(Button)>
</form>
`;

exports[`Test within Form shallow a TextInput 1`] = `
<form>
  <TextInput
    doc={
      Object {
        "test": "Type Sample",
      }
    }
    fieldName="test"
    key=".0"
    label="test"
    setDoc={[Function]}
    value="Type Sample"
  />
  <WithStyles(Button)
    onClick={[Function]}
  >
    Submit
  </WithStyles(Button)>
</form>
`;
